partner_affiliate/models/res_partner.py

```diff
@@ -1,24 +1,156 @@
+    #             return super().write({"type": "contact"})
+    #     return super().onchange_parent_id()
+
+    @api.onchange("city", "parent_id", "company_type")
+    def _onchange_city(self):
+        if self.type == "affiliate":
+            self.get_affiliate_name()
+
+    def get_affiliate_name(self):
+        parent_name = self.parent_id.name.upper() if self.parent_id.name else ""
+        city_name = self.city.upper() if self.city else ""
+        if parent_name and city_name:
+            self.name = f"{parent_name} - {city_name}"
+        else:
+            self.name = parent_name or city_name
+    
+    def _fields_sync(self, values):
+        """A company that will have a parent set, must be an affiliate,
+
+        Setting a company to type affiliate, will automatically prevent
+        overriding the address from the parent, because that is only done for
+        contact partners.
+
+        Clearing a parent_id from a company should reset its type to contact.
+
+        You could argue that the same should be true for individuals.
+        """
+        if self.company_type == "person":
+            desired_type = "contact"
+        elif not self.parent_id and self.company_type == "company":
+            desired_type = "parent_company"
+        elif self.parent_id and self.company_type == "company":
+            desired_type = "affiliate"
+        else:
+           desired_type = "contact"
+        
+        if self.type != desired_type:
+            super().write({"type": desired_type})
+        return super()._fields_sync(values)
+
+    @api.model
+    def create(self, values):
+        res = super().create(values)
+        self.check_if_parent_company(values)
+        if res.parent_id:
+            res.vat = None
+            return res
+        return res
+
+    def check_if_parent_company(self, values):
+        if values.get("type") == "parent_company":
+            if not values.get("vat"):
+                raise ValidationError(_("Vat is mandatory to create a Parent Company"))
+
+    # @api.depends("company_type")
+    # def _compute_parent_id_domain(self):
+    #     for partner in self:
+    #         if partner.company_type == "person":
+    #             partner.parent_id_domain = ['affiliate', 'parent_company']
+    #         else:
+    #             partner.parent_id_domain = ['parent_company']
+
+    @api.onchange("company_type")
+    def onchange_get_partner_ids_domain(self):
+        if self.company_type == 'person':
+            domain = [('type','in', ['affiliate', 'parent_company'])]
+        else:
+            domain = [('type','in', ['parent_company'])]
+        return {'domain': {'parent_id': domain}}
+
+    @api.depends("company_type")
+    def get_partner_ids_domain(self):
+        if self.company_type == 'person':
+            domain = [('type','in', ['affiliate', 'parent_company'])]
+        else:
+            domain = [('type','in', ['parent_company'])]
+        return {'domain': {'parent_id': domain}}
```

Comment:
è un recorset e non puoi utilizzare il self in questo modo, stesso discorso per il return